Add support for materialized CTEs

kryonix opened the pull request:
This pr adds support for materialized CTE expressions as an alternative to inlining, as I briefly mentioned in #404.

Inlining CTEs can be great, but it can also multiply the work. Take this query for example:

```
WITH t(x) AS
  (SELECT t FROM generate_series(1,10) AS _(t))
SELECT t1.x, t2.x FROM t AS t1, t AS t2;
```

Inlining duplicates the definition of `t` for each reference.
This results in the following query:

```
SELECT t1.x, t2.x FROM (SELECT t FROM generate_series(1,10) AS _(t)) AS t1(x),
                       (SELECT t FROM generate_series(1,10) AS _(t)) AS t2(x);
```

Now, if `t` is an expensive query, there may be a problem.

---

With this pr, CTEs can be annotated as being `MATERIALIZED`.
Such CTEs first fully materialize their result, which can then be read:

```
WITH t(x) AS MATERIALIZED
  (SELECT t FROM generate_series(1,10) AS _(t))
SELECT t1.x, t2.x FROM t AS t1, t AS t2;
```

Instead of copying `t` for each reference, `t` is calculated exactly once. Since the result is cached, no work is done multiple times.

kryonix committed:
Add parsing for MATERIALIZED CTEs

kryonix committed:
Add new CTENode for materialized CTEs to parser

kryonix committed:
Fix transformation of materialized CTEs for recursive CTEs

kryonix committed:
Add bound_cte_node to bind materialized CTEs

kryonix committed:
Add binding of materialized CTEs

kryonix committed:
Add LOGICAL_CTE operator and type

kryonix committed:
Add logical plan generation for materialized CTEs

kryonix committed:
Fix comment in BaseTableRef binder

kryonix committed:
Add physical operator types CTE and CTE_SCAN

kryonix committed:
Add cte_dependencies to Pipeline class to keep track of CTE_SCAN operations

kryonix committed:
Add materialized_ctes and materialized_cte_ids to physical plan generator

kryonix committed:
Add PhysicalCTE operator

kryonix committed:
Add physical plan generation for LogicalCTE node

kryonix committed:
Adapt physical plan generation of LogicalCTERef for materialized CTEs

kryonix committed:
Adapt BuildPipelines of physical_column_data_scan for materialized CTEs

kryonix committed:
Move materialized CTE handling to Transformer

kryonix committed:
Improve TransformSelectInternal and add CTE materialization for set operations

kryonix committed:
Throw exception when materialized CTE is used in DELETE statement

kryonix committed:
Throw exception when materialized CTE is used in INSERT statement

kryonix committed:
Throw exception when materialized CTE is used in UPDATE statement

kryonix committed:
Throw exception when materialized CTE is used in PIVOT statement

kryonix committed:
Fix toString method of CTENode

kryonix committed:
Remove unused assertion from LogicalCTERef physical plan generator

kryonix committed:
Improve EXPLAIN output of CTE, RECURSIVE_CTE and COLUMN_DATA_SCAN

kryonix committed:
Add test cases for materialized CTEs

kryonix committed:
Add check for unplanned subqueries to logical planning of BoundCTENode

kryonix commented:
This pr is currently a draft, as there are two remaining issues:

---

First, something during binding seems to be broken, as the following fails:
```
SELECT x FROM (WITH t(y, z) AS MATERIALIZED (SELECT 1, 2) SELECT * FROM t) AS _(x, y);
```

```
Error: INTERNAL Error: Failed to bind column reference "x" [0.0] (bindings: [9.0])
```

---

Second, pipeline construction of two (or more) materialized recursive CTEs fails:
```
WITH RECURSIVE t(x) AS MATERIALIZED
(
  SELECT 1
    UNION ALL
  SELECT x+1
  FROM   t
  WHERE  x < 4
)
,
u(x) AS
(
  SELECT *
  FROM   t
    UNION ALL
  SELECT u.x + t.x
  FROM   u, t
  WHERE  u.x < 32
)
SELECT *
FROM   u;
```

```
Error: INTERNAL Error: Assertion triggered in file "/Users/denis/Documents/git/duckdb/src/parallel/executor.cpp" on line 156: event_map_entry != event_map.end()
```

kryonix committed:
Add CTE_NODE to statement simplifier in sqlsmith

kryonix committed:
Fix format

kryonix committed:
Fix after rebase

Mytherin commented on src/planner/binder/query_node/bind_cte_node.cpp:
```diff
@@ -0,0 +1,64 @@
+	statement.modifiers.clear();
+
+	// Add bindings of left side to temporary CTE bindings context
+	result->child_binder->bind_context.AddCTEBinding(result->setop_index, statement.ctename, result->names,
```
Is the correct table index used here (`result->setop_index`)? This might be where the column binding error comes from.

Mytherin reviewed (commented):
Thanks for the PR! Looks good - very cool feature! I have left some comments inline. One note is that we have been working to remove C-style pointers (see #7080) - this might cause a few merge conflicts on your end. 

> SELECT x FROM (WITH t(y, z) AS MATERIALIZED (SELECT 1, 2) SELECT * FROM t) AS _(x, y);

> Error: INTERNAL Error: Failed to bind column reference "x" [0.0] (bindings: [9.0])


This is the column binding resolver that complains. Essentially "x" is bound as the first column to table binding `0`, but table binding `0` is not accessible from that part of the plan.

> Second, pipeline construction of two (or more) materialized recursive CTEs fails:

I'm not sure about what causes this exactly - I would guess that something goes wrong in the order in which `BuildPipelines` is executed so that the layered situation is not resolved correctly, or perhaps the incorrect pipeline is passed on to the child.

Perhaps it would also be interesting to add a benchmark where the benefits of the materialized CTEs are visible?

kryonix committed:
Fix deserialization of CTEMaterialize field of CTE Info

kryonix committed:
Adapt recursive cte tutorial materialized test as LIMIT cannot stop iteration in that case

kryonix committed:
Prevent RemoveUnusedColumns optimizer to remove LogicalCTE columns

kryonix committed:
Add column_count to LogicalCTE just in case

kryonix committed:
Fix GetRootIndex of BoundCTE

kryonix committed:
Add some missing serialization and deserializations for CTE map

kryonix committed:
Add flag to CTE nodes which indicates if it is materialized

kryonix committed:
Remove materialized_cte_ids from planner and use materialized flags instead

kryonix commented on src/execution/physical_plan/plan_recursive_cte.cpp:
```diff
@@ -30,8 +30,31 @@ unique_ptr<PhysicalOperator> PhysicalPlanGenerator::CreatePlan(LogicalRecursiveC
-	    make_uniq<PhysicalColumnDataScan>(op.types, PhysicalOperatorType::RECURSIVE_CTE_SCAN, op.estimated_cardinality);
+	// Check if this LogicalCTERef is supposed to scan a materialized CTE.
+	for (idx_t i = 0; i < materialized_cte_ids.size(); i++) {
+		if (op.cte_index == materialized_cte_ids[i]) {
```
`materialized_cte_ids` are no longer required.

kryonix commented on src/include/duckdb/execution/physical_plan_generator.hpp:
```diff
@@ -32,6 +32,9 @@ class PhysicalPlanGenerator {
 	unordered_map<idx_t, std::shared_ptr<ColumnDataCollection>> recursive_cte_tables;
+	//! Materialized CTE scans need to be collected for each CTE operator.
+	unordered_map<idx_t, vector<PhysicalOperator *>> materialized_ctes;
+	vector<idx_t> materialized_cte_ids;
```
I found a better way to do this. This indeed allows to perform the lookup in `materialized_ctes`.

kryonix commented:
The binder issue seems to be resolved now.

Thank you for the comments. I will address them.

> Thanks for the PR! Looks good - very cool feature! I have left some comments inline. One note is that we have been working to remove C-style pointers (see https://github.com/duckdb/duckdb/pull/7080) - this might cause a few merge conflicts on your end.

I think I will just rebase onto master and do the rewritings before I take this pr out of its draft status. Should be a clean solution, I think.

kryonix committed:
Reset materialized result when the sink resets

kryonix committed:
Add test for materialized CTE in recursive CTE

kryonix committed:
Remove C-style pointers from MATERIALIZED CTEs

kryonix committed:
Fix format

kryonix committed:
Rework MATERIALIZED CTE pipelines to work similar to RECURSIVE CTEs

kryonix committed:
Adapt physical_cte operator to new PhysicalOperator interface

kryonix committed:
Format fix

kryonix committed:
Adapt testcase

kryonix committed:
Fix tidy-check

kryonix committed:
Add benchmark files for materialized CTE

kryonix commented:
I was able to fix the pipelining for materialized CTEs. However, I had to adapt an execution model similar to the one used by recursive CTEs. That's because previously a pipeline dependency `CTE_SCAN 0 -> p₁` was used to ensure correct execution order. Just like delim joins do. This does not work if `p₂` is a recursive CTE, because then `p₁` is not part of that executor. So the scheduling fails.

```
CTE 0
|   \
|    \
p₁   p₂
     ⋮
     CTE_SCAN 0
```

kryonix committed:
Fix tidy-check

kryonix committed:
Rename identifiers in executor to match materialized CTEs

Mytherin commented:
Thanks for the fixes! I will review the PR next week - we are doing a release on Monday so this week is rather packed :)

Mytherin commented on src/include/duckdb/planner/tableref/bound_cteref.hpp:
```diff
@@ -29,5 +30,7 @@ class BoundCTERef : public BoundTableRef {
 	//! The index of the cte
 	idx_t cte_index;
+	//! Is this a reference to a materialized CTE?
+	bool materialized_cte;
```
Can we use the `CTEMaterialize` enum here?

kryonix replied:
Yes, `CTEMaterialize` is used now.

Mytherin commented on src/parser/query_node.cpp:
```diff
@@ -230,6 +243,7 @@ unique_ptr<QueryNode> QueryNode::Deserialize(Deserializer &main_source) {
 		auto info = make_uniq<CommonTableExpressionInfo>();
 		source.ReadStringVector(info->aliases);
 		info->query = SelectStatement::Deserialize(source);
+		info->materialized = reader.ReadRequired<CTEMaterialize>();
```
Can we use `ReadField` here to avoid breaking the storage version?

kryonix replied:
Done.

kryonix replied:
This did not solve the issue. I tried debugging this, but the error seems to not occur in any code I touched. Maybe you have an idea based on the callstack?

```
    frame #0: 0x0000000121bc5c74 libduckdb.dylib`duckdb::FieldDeserializer::ReadData(this=0x000000016fdf51d0, buffer="\U0000000e6\xe0E", read_size=8) at field_writer.cpp:57:3
   54  	void FieldDeserializer::ReadData(data_ptr_t buffer, idx_t read_size) {
   55  		D_ASSERT(remaining_data != idx_t(-1));
   56  		if(read_size <= remaining_data) {
-> 57  			throw Exception("error");
   58  		}
   59  		D_ASSERT(read_size <= remaining_data);
   60  		root.ReadData(buffer, read_size);
(lldb) up
frame #1: 0x0000000100c22e1c unittest`unsigned long long duckdb::Deserializer::Read<unsigned long long>(this=0x000000016fdf51d0) at serializer.hpp:130:3
   127 		template <class T>
   128 		T Read() {
   129 			T value;
-> 130 			ReadData((data_ptr_t)&value, sizeof(T));
   131 			return value;
   132 		}
   133
(lldb) up
frame #2: 0x000000011c2b495c libduckdb.dylib`unsigned long long duckdb::FieldReader::ReadRequired<unsigned long long>(this=0x000000016fdf51d0) at field_writer.hpp:200:17
   197 			}
   198 			// field is there, read the actual value
   199 			AddField();
-> 200 			return source.Read<T>();
   201 		}
   202
   203 		template <class T>
(lldb) up
frame #3: 0x0000000124789f04 libduckdb.dylib`duckdb::MainHeader::Deserialize(source=0x000000016fdf54e0) at single_file_block_manager.cpp:76:28
   73  		// read the flags
   74  		FieldReader reader(source);
   75  		for (idx_t i = 0; i < FLAG_COUNT; i++) {
-> 76  			header.flags[i] = reader.ReadRequired<uint64_t>();
   77  		}
   78  		reader.Finalize();
   79  		return header;
(lldb) up
frame #4: 0x000000012478d8a0 libduckdb.dylib`duckdb::MainHeader duckdb::DeserializeHeaderStructure<duckdb::MainHeader>(ptr="DUCK3") at single_file_block_manager.cpp:107:9
   104 	template <class T>
   105 	T DeserializeHeaderStructure(data_ptr_t ptr) {
   106 		BufferedDeserializer source(ptr, Storage::FILE_HEADER_SIZE);
-> 107 		return T::Deserialize(source);
   108 	}
   109
   110 	SingleFileBlockManager::SingleFileBlockManager(AttachedDatabase &db, string path_p, StorageManagerOptions options)
(lldb) up
frame #5: 0x000000012474cefc libduckdb.dylib`duckdb::SingleFileBlockManager::LoadExistingDatabase(this=0x0000000106141080) at single_file_block_manager.cpp:194:2
   191 		MainHeader::CheckMagicBytes(*handle);
   192 		// otherwise, we check the metadata of the file
   193 		ReadAndChecksum(header_buffer, 0);
-> 194 		DeserializeHeaderStructure<MainHeader>(header_buffer.buffer);
   195
   196 		// read the database headers from disk
   197 		DatabaseHeader h1, h2;
(lldb) up
frame #6: 0x0000000124748828 libduckdb.dylib`duckdb::SingleFileStorageManager::LoadDatabase(this=0x0000000105096ac0) at storage_manager.cpp:125:21
   122 		} else {
   123 			// initialize the block manager while loading the current db file
   124 			auto sf_block_manager = make_uniq<SingleFileBlockManager>(db, path, options);
-> 125 			sf_block_manager->LoadExistingDatabase();
   126 			block_manager = std::move(sf_block_manager);
   127 			table_io_manager = make_uniq<SingleFileTableIOManager>(*block_manager);
   128
(lldb) up
frame #7: 0x0000000124746110 libduckdb.dylib`duckdb::StorageManager::Initialize(this=0x0000000105096ac0) at storage_manager.cpp:66:2
   63  		}
   64
   65  		// create or load the database from disk, if not in-memory mode
-> 66  		LoadDatabase();
   67  	}
   68
   69  	///////////////////////////////////////////////////////////////////////////
 ```

kryonix replied:
Changing this to `ReadField` does not fix the issue. It works if I remove the line completely. But that's not a solution.

Mytherin replied:
Indeed you are right - the issue here is that a `FieldWriter` is used for all CTEs, which essentially disallows adding fields to the end of individual CTEs. Could you increment the storage version and re-generate by running (`scripts/generate_storage_version.py`)?

Mytherin commented on src/planner/operator/logical_cte.cpp:
```diff
@@ -0,0 +1,19 @@
+
+namespace duckdb {
+
+void LogicalCTE::Serialize(FieldWriter &writer) const {
```
The `LogicalCTE` is only created for materialized CTEs, right? Could we perhaps rename it to `LogicalMaterializedCTE` for more clarity?

kryonix replied:
Makes sense. It is renamed now.

Mytherin commented on src/planner/binder/tableref/bind_basetableref.cpp:
```diff
@@ -45,9 +45,11 @@ unique_ptr<BoundTableRef> Binder::Bind(BaseTableRef &ref) {
 			auto index = GenerateTableIndex();
-			auto result = make_uniq<BoundCTERef>(index, ctebinding->index);
+			auto result = make_uniq<BoundCTERef>(index, ctebinding->index,
+			                                     cte.materialized == CTEMaterialize::CTE_MATERIALIZE_ALWAYS);
```
As materialized and non-materialized CTEs should always have the same result, perhaps it makes sense to switch over the default when alternative verification is enabled. For example:

```cpp
auto materialized = cte.materialized;
if (materialized == CTEMaterialize::CTE_MATERIALIZE_DEFAULT) {
#ifdef DUCKDB_ALTERNATIVE_VERIFY
   materialized = CTEMaterialize::CTE_MATERIALIZE_ALWAYS;
#else
   materialized = CTEMaterialize::CTE_MATERIALIZE_NEVER;
#endif
}
...
```

That way when compiling with `ALTERNATIVE_VERIFY=1 make` all CTEs will be materialized by default. This will enable some free extra testing.

kryonix replied:
In principle, you are right—and I like the idea. However, there is a corner case, where `MATERIALIZED` CTEs behave differently:

```
WITH RECURSIVE t(x) AS (
  SELECT 1
    UNION ALL
  SELECT x+1
  FROM   t
)
SELECT * FROM t LIMIT 4;
```

This query expresses an infinite loop. It only stops, because of the `LIMIT` clause.
If you materialize this CTE, it does *not* terminate anymore.

Mytherin replied:
That makes sense. I imagine there are only a few tests where those types of CTEs are run, though. Perhaps we could add a new option (e.g. `noalternateverify`) to those tests - and add `require noalternateverify` to them so the materialized CTE test is skipped?

kryonix replied:
Yes, there are a few tests, but not many. Adding such an option sounds good to me. I will look into it.

kryonix replied:
I've added the new option `noalternativeverify` to the unittest and annotated the relevant tests with it.

Mytherin reviewed (commented):
Thanks for the fixes! Looks great. Some more comments below. I have switched the merge target to the feature branch as we are only merging bug-fixes into master for now. 

There is a failing test case related to the storage version which should be fixed by the change mentioned in the comments.

The merge conflict should be resolved by merging with feature and re-running the generate_grammar script.

kryonix committed:
Rename logical CTE to logical materialized CTE

Fix storage version

Use CTEMaterialized enum instead of boolean

kryonix committed:
Add new option 'noalternativeverify' to unittest to skip certain tests

kryonix committed:
Annotate infinite loop recursive CTE test cases with noalternativeverify

kryonix committed:
Invert MATERIALIZED flag when alternative verify is used

kryonix committed:
Merge remote-tracking branch 'origin/feature'

kryonix committed:
Fix format

kryonix committed:
Fix storage version

kryonix committed:
Move CTEMaterialize enum to common/enums/cte_materialize.hpp

kryonix committed:
Merge remote-tracking branch 'origin/feature'

kryonix committed:
Fix plan_cte_node.cpp

Mytherin commented:
Could you also have a look at the code coverage, minus the issues addressed by #7636 (see #7611)?

kryonix committed:
Remove TransformCTE and rename TransformCTEInternal to TransformCTE

kryonix committed:
Increment storage version number to make it work for materialized CTEs

kryonix committed:
Add limit percent test for materialized CTEs

kryonix committed:
Fix format

kryonix committed:
Add materialized CTE nodes to enum_util.cpp

kryonix committed:
Add some expected lines to uncovered_files for materialized CTEs

kryonix committed:
Merge remote-tracking branch 'origin/feature'

kryonix committed:
Enable materialized CTEs for PIVOT statements and add tests for it

kryonix commented:
I think all remaining issues are now resolved.

kryonix committed:
Fix alternative verify for materialized CTEs

Some test cases can (and should!) be re-enabled when
decorrelation of materialized CTEs is implemented.

kryonix committed:
Fix two more alternative verify errors

kryonix committed:
Fix code coverage

kryonix committed:
Throw binder error when trying to flatten correlated materialized CTE

kryonix committed:
Fix uncovered_files.csv

kryonix committed:
Fix more alternative verify issues

kryonix committed:
Merge remote-tracking branch 'origin/feature'

kryonix committed:
Add tests for materialized CTEs and DML statements

kryonix committed:
Merge remote-tracking branch 'origin/feature'

kryonix committed:
Fix materialized cte DML tests

kryonix committed:
Add noalternativeverify to materialized dml CTEs

kryonix committed:
Only skip tests with require alternativeverify, when in alternative verify mode

kryonix commented:
If you find the time, @Mytherin, I think this pr is ready for review now.

Mytherin commented on src/parser/transform/statement/transform_update.cpp:
```diff
@@ -19,8 +19,13 @@ unique_ptr<UpdateSetInfo> Transformer::TransformUpdateSetInfo(duckdb_libpgquery:
+		TransformCTE(*PGPointerCast<duckdb_libpgquery::PGWithClause>(stmt.withClause), result->cte_map,
+		             materialized_ctes);
+		if (!materialized_ctes.empty()) {
+			throw NotImplementedException("Materialized CTEs are not implemented for delete.");
```
Should this be `update` - also could you perhaps expand on the reason for this?

kryonix replied:
You are right, the message is wrong. I will fix that.

---

There is an AST 'mismatch' currently. Materialized CTEs are represented as a `QueryNode`, whereas`INSERT/UPDATE/DELETE` are `SQLStatements`. This means, that it is easy to represent the materialized CTEs as a bunch of stacked `QueryNodes` (this happens in `Transformer::TransformMaterializedCTE`). It is not that easy for `INSERT/UPDATE/DELETE`, however. While there should not be a principle problem implementing that, my current plan is to tackle that in a separate pr.

Mytherin commented on test/sql/cte/materialized/dml_materialized_cte.test:
```diff
@@ -0,0 +1,26 @@
+statement ok
+PRAGMA enable_verification
+
+require noalternativeverify
```
Is this necessary here?

kryonix replied:
Currently, yes. Until `MATERIALIZED` CTEs are implemented for `UPDATE/INSERT/DELETE`-statements, alternative verify would lead to non-`MATERIALIZED` CTEs here. Therefore, the tests would be unexpectedly successful.

Mytherin reviewed (commented):
Thanks for the fixes! Looks great to me now. One minor comment otherwise this is good to go. Could you also merge with `feature` so the CI can rerun? I believe the CI failures should be fixed then.

kryonix committed:
Fix materialized CTE exception in transform_update.cpp

kryonix committed:
Merge remote-tracking branch 'origin/feature'

kryonix committed:
Merge remote-tracking branch 'origin/feature'

Mytherin committed:
Fix uncovered files

Mytherin commented:
Thanks!
